[Merged by Bors] - feat(Tactic/CategoryTheory): normalization in monoidal categories

yuma-mizuno opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vtec234 commented on Mathlib/Tactic/CategoryTheory/Monoidal.lean:
```diff
@@ -0,0 +1,681 @@
+
+open Mathlib.Tactic.Monoidal
+
+/-- `normalize% η` is the normalization of the 2-morphism `η`. It is of the form
+`α₀ ≫ η₀ ≫ α₁ ≫ η₁ ≫ ... αₙ ≫ ηₙ ≫ αₙ₊₁`, where `αᵢ` are structural 2-morphisms
+and `ηᵢ` are non-structural 2-morphisms. -/
```
This description of normal forms seems a bit less specific than the description in the module doc at the top. Does the more specific description apply here?

Vtec234 commented:
What is the status of this PR? Is it still a draft or could be marked `awaiting-review`?

yuma-mizuno commented:
@Vtec234 Thank you for the review! I think it's ready for review. I've opened the PR.

BoltonBailey reviewed (approved):
I looked over this, and nothing jumped out, though I am not super experienced with metaprogramming code.

Perhaps @yuma-mizuno could either resolve the above conversation or mark this PR as awaiting author?

yuma-mizuno commented:
I apologize for the late response. I updated the docs following the suggestion by @Vtec234's suggestion.

I also added a note about what to do if proof term generation fails. The failure to generate proof terms occurs when using the normalization function in the string diagram widget. As a workaround we generate a meaningless term instead of generating an error.

semorrison commented on Mathlib/Tactic/CategoryTheory/Monoidal.lean:
```diff
@@ -0,0 +1,698 @@
+3. `θ` is of the form `ι ▷ g₁ ▷ ... ▷ gₗ`
+
+Note that the structural morphisms `αᵢ` are not necessarily normalized, as the main purpose
+to be to get a list of the non-structural morphisms out.
```
```suggestion
is to get a list of the non-structural morphisms out.
```

semorrison commented:
Could you write an account somewhere of your plans with this (or point to existing Zulip discussion etc)?

I'm excited to have widget support for string diagrams, but I would like to know that
* this code will be maintainable into the future
* its learnt whatever lessons necessary from e.g. homotopy.io

and I think a bit more documentation would help produce this confidence!

yuma-mizuno commented:
@semorrison Thanks for the comment! I have added the note explaining the current situation and the future plan.

I've played a bit with homotopy.io before, but to be honest I don't know much about it (But now I should learn it).

semorrison commented on Mathlib/Tactic/CategoryTheory/Monoidal.lean:
```diff
@@ -0,0 +1,726 @@
+normalization procedure in this file does not rely on specific settings, allowing for broader
+application.
+
+Future plans include the following. At least I (Yuma) would like to work on tehse in the future,
```
```suggestion
Future plans include the following. At least I (Yuma) would like to work on these in the future,
```

semorrison commented on Mathlib/Tactic/CategoryTheory/Monoidal.lean:
```diff
@@ -0,0 +1,726 @@
+    catch _ => return ⟨ι'', mkConst ``True⟩
+  | .id, η => do
+    let f := η.src
+    let g := η.tar
```
I think `tgt` is a better abbreviation.

semorrison commented:
bors d+

yuma-mizuno commented:
bors r+
